Riot/Modules/TabBar/TabBarCoordinator.swift

```diff
@@ -699,8 +742,190 @@ final class TabBarCoordinator: NSObject, TabBarCoordinatorType {
+            avatarView.isUserInteractionEnabled = false
+            avatarView.fill(with: avatar)
+            avatarView.update(theme: ThemeService.shared().theme)
+            avatarView.autoresizingMask = [.flexibleHeight, .flexibleWidth]
+            view.addSubview(avatarView)
+            self.rightMenuAvatarView = avatarView
+        }
+        
+        self.masterTabBarController.navigationItem.rightBarButtonItem = UIBarButtonItem(customView: view)
+    }
+    
+    // MARK: Sign out process
+    
+    private func signOut() {
+        guard let keyBackup = currentMatrixSession?.crypto.backup else {
+            return
+        }
+        
+        signOutAlertPresenter.present(for: keyBackup.state,
+                                      areThereKeysToBackup: keyBackup.hasKeysToBackup,
+                                      from: self.masterTabBarController,
+                                      sourceView: nil,
+                                      animated: true)
+    }
+    
+    // MARK: - SecureBackupSetupCoordinatorBridgePresenter
+    
+    private var secureBackupSetupCoordinatorBridgePresenter: SecureBackupSetupCoordinatorBridgePresenter?
+    private var crossSigningSetupCoordinatorBridgePresenter: CrossSigningSetupCoordinatorBridgePresenter?
+
+    private func showSecureBackupSetupFromSignOutFlow() {
+        if canSetupSecureBackup {
+            setupSecureBackup2()
+        } else {
+            // Set up cross-signing first
+            setupCrossSigning(title: VectorL10n.secureKeyBackupSetupIntroTitle,
+                              message: VectorL10n.securitySettingsUserPasswordDescription) { [weak self] result in
+                switch result {
+                case .success(let isCompleted):
+                    if isCompleted {
+                        self?.setupSecureBackup2()
+                    }
+                case .failure:
+                    break
+                }
+            }
+        }
+    }
+    
+    private var canSetupSecureBackup: Bool {
+        // Accept to create a setup only if we have the 3 cross-signing keys
+        // This is the path to have a sane state
+        // TODO: What about missing MSK that was not gossiped before?
+        
+        guard let recoveryService = currentMatrixSession?.crypto.recoveryService else {
+            return false
+        }
+        
+        let crossSigningServiceSecrets = [
+            MXSecretId.crossSigningMaster.takeRetainedValue() as String,
+            MXSecretId.crossSigningSelfSigning.takeRetainedValue() as String,
+            MXSecretId.crossSigningUserSigning.takeRetainedValue() as String
+        ]
+        
+        return recoveryService.secretsStoredLocally().filter { secret in
+            guard let secret = secret as? String else { return false }
+            return crossSigningServiceSecrets.contains(secret)
+        }.count == crossSigningServiceSecrets.count
```

Comment:
I wonder if this little chunk of code could be make into a method on the recoveryService so it can be called from both here and the `SettingsViewController` without duplication?

Reply:
Just found `vc_canSetupSecureBackup` method in `MXSession` extension 😩